Adding fixed hash length

LukePearson1 opened the pull request:
This PR adds a function which allows us to take in one BlsScalar and output 2. This is for use in the EdDSA scheme for JubJub. 
This closes #51.

Commit:
Adding fixed hash length

For some cases the output of the
sponge hash needs to vary. This
implementation builds a hash function
with input 1 and output 2.

CPerezz reviewed (changes requested):
The impl has nothing to do with the specs mentioned in https://app.gitbook.com/@dusk-network/s/specs/specifications/poseidon/poseidon-usages/

Please, check fixed-len hashing section. 
I can go through it with you if you want.

Commit:
Include mod.rs

mod.rs is included in this commit
so the tests can be included, which
are passing

Commit:
Add perm uses file

Varying uses of permuatations on
the poseidon hash function are to
be included.

This file adds the fixed hash for
two outputs of the same input.

Commit:
Refactor sponge file

The sponge file removes the fixed hash
and places it into the use cases file

Commit:
Change test name

Take out 'sponge' from name of test
as the file has changed away from the
sponge folder.

CPerezz commented on src/sponge/pad.rs:
```diff
@@ -35,6 +35,26 @@ where
+    words[1] = message;
+    words
+}
+
```
This pad should not be inside the sponge module. 
Please move it under the same module where the fixed hash file lives.

CPerezz commented on src/perm_uses/fixed_hash.rs:
```diff
@@ -0,0 +1,71 @@
+        let h = two_outputs(m); 
+        let h_1 = two_outputs(m);
+        
+        assert_eq!(h, h_1);
```
Could you wrap this inside of a loop that does the operation 100 times or so? On that way we are sure that it doesn't fail.

CPerezz commented on src/perm_uses/fixed_hash.rs:
```diff
@@ -0,0 +1,71 @@
+            strategy.perm(&mut inputs);
+            inputs
+        },
+    );
```
Why do you need to do any of these things?

You already know that the inputs are as max: WIDTH - 1. Therefore it makes no sense that you have here code copied from the sponge that doesn't have anything to do and that will add functionalities that you don't want.

LukePearson1 replied:
I have refactored this for use with fixed width.

CPerezz commented on src/perm_uses/fixed_hash.rs:
```diff
@@ -0,0 +1,71 @@
+    let padder = BlsScalar::zero();
+
+    // The capacity is 
+    let capacity = BlsScalar::one() * BlsScalar::from(2<<64-1) + BlsScalar::one(); 
```
This could be a constant.

LukePearson1 replied:
I don't want to encode it as a constant for now, incase we opt for the change in number of inputs and outputs.

CPerezz replied:
Is there any reason why not? It will be faster and cleaner. The only reason to have that is if you were doing a generic impl over the ammount of inputs/outputs

LukePearson1 replied:
Because if the use case of the fixed hash is going to become something other EdDSA, then we cannot have capacity as constant. Linked to the comment below, if we have an in circuit use, in the future, of a fixed hash then both a generic function and varying `capacity` will be needed.

CPerezz commented on src/sponge/pad.rs:
```diff
@@ -35,6 +35,26 @@ where
+    // For a constant length hash, we always use a slice of width 5.
+    let width = 5;
+    let zero = pad_value;
+    let mut words = vec![zero; width];
```
As I said, would not be difficult to turn that into a generic function for any number of inputs/outputs < WIDTH.

Anyway, for the time being maybe that's fine. Even I encourage you to try it.

LukePearson1 replied:
I would like to do this in the future when it is optimised for circuits.

CPerezz replied:
Doesn't have anything to do with circuits but ok

LukePearson1 replied:
See above comment for my train of thought

Commit:
Refactor to review comments

A review on this code has lead to
the refactoring of the file.

This includes: Removing certain
functionality as width is fixed.
Ierating a test to check more
values.

CPerezz reviewed (approved):
LGTM
